shared/core/src/main/scala/org/hyperledger/identus/shared/messaging/kafka/InMemoryMessagingService.scala

```diff
@@ -0,0 +1,169 @@
+              ZIO.succeed(ZStream.fromQueue(queue).collect { case msg: Message[K, V] @unchecked => (topic, msg) })
+          case None =>
+            ZIO.sleep(1.second) *> ZIO.succeed(ZStream.empty)
+        }
+      }.flatten
+
+    val streams = allTopics.map(getQueueStream)
+    ZStream
+      .mergeAllUnbounded()(streams: _*)
+      .tap { case (topic, msg) => ZIO.log(s"Processing message in group $groupId, topic:$topic : $msg") }
+      .filterZIO { case (topic, msg) =>
+        for {
+          currentTime <- Clock.currentTime(TimeUnit.MILLISECONDS)
+          key = ConsumerGroupKey(groupId, topic)
+          topicProcessedMessages <- processedMessagesMap.get(key).flatMap {
+            case Some(map) => ZIO.succeed(map)
+            case None =>
+              for {
+                newMap <- ConcurrentMap.empty[Offset, TimeStamp]
+                _ <- processedMessagesMap.put(key, newMap)
+              } yield newMap
+          }
+          isNew <- topicProcessedMessages
+            .putIfAbsent(Offset(msg.offset), TimeStamp(currentTime))
+            .map(_.isEmpty)
+        } yield isNew
+      }
+      .mapZIO { case (_, msg) => handler(msg) }
+      .tap(_ => ZIO.log(s"Message processed in group $groupId, topic:$topic"))
+      .runDrain
+  }
+}
+
+class InMemoryProducer[K, V](
+    topicQueues: ConcurrentMap[Topic, (Queue[Message[_, _]], Ref[Offset])],
+    queueCapacity: Int
+) extends Producer[K, V] {
+  override def produce(topic: String, key: K, value: V): Task[Unit] = for {
+    queueAndOffsetRef <- topicQueues.get(topic).flatMap {
+      case Some(qAndOffSetRef) => ZIO.succeed(qAndOffSetRef)
+      case None =>
+        for {
+          newQueue <- Queue.sliding[Message[_, _]](queueCapacity)
+          newOffSetRef <- Ref.make(Offset(0L))
+          _ <- topicQueues.put(topic, (newQueue, newOffSetRef))
+        } yield (newQueue, newOffSetRef)
+    }
+    (queue, offsetRef) = queueAndOffsetRef
+    currentTime <- Clock.currentTime(TimeUnit.MILLISECONDS)
+    messageId <- offsetRef.updateAndGet(x => Offset(x.value + 1)) //  unique atomic id incremented per topic
+    _ <- queue.offer(Message(key, value, messageId.value, currentTime))
+  } yield ()
+}
+
+object InMemoryMessagingService {
+  type Topic = String
+  type GroupId = String
+
+  opaque type Offset = Long
+  object Offset:
+    def apply(value: Long): Offset = value
+    extension (id: Offset) def value: Long = id
+
+  opaque type TimeStamp = Long
+  object TimeStamp:
+    def apply(value: Long): TimeStamp = value
+    extension (ts: TimeStamp) def value: Long = ts
+
+  val messagingServiceLayer: ULayer[MessagingService] =
+    ZLayer.fromZIO {
+      for {
+        queueMap <- ConcurrentMap.empty[Topic, (Queue[Message[_, _]], Ref[Offset])]
+        processedMessagesMap <- ConcurrentMap.empty[ConsumerGroupKey, ConcurrentMap[Offset, TimeStamp]]
+        queueCapacity = 100 // queue capacity
```

Comment:
Should it be configurable?
100 might be not enough

Reply:
Done 😉